### ISSUE TITLE
Numerical grad inconsistencies: grad * op_backward or directly op_backward

### ISSUE THREAD

**Issue by mratsim** [mratsim/Arraymancer#142]
Convolution:
https://github.com/mratsim/Arraymancer/blob/69f3c32b7a375092bba869100aea23c48d06232e/tests/nn_primitives/test_nnp_convolution.nim#L153-L158

Softmax cross-entropy:
https://github.com/mratsim/Arraymancer/blob/69f3c32b7a375092bba869100aea23c48d06232e/tests/nn_primitives/test_nnp_loss.nim#L39-L47

Not sure if it's normal or not.

**Comment by mratsim** [comment]
New case with maxpool:

https://github.com/mratsim/Arraymancer/blob/30bba6761eebc5074dbb16a5e6a97173aff201a2/tests/nn_primitives/test_nnp_maxpool.nim#L32-L41

Numerical gradients only returns the 1 or 0.

**Comment by sherjilozair** [comment]
Could you explain what you mean by this issue? The tests pass for me on CPU. Do they fail on GPU?

**Comment by mratsim** [comment]
It's more like for some tests I have to multiply by the gradient flowing back i.e.
`let expected = grad .* op_backward`

And for others I don't: 
`let expected = op_backward` directly.

I'm pretty sure there is an explanation and no error but I'm curious about this inconsistency.
